src/storage/hummock_memory_limiter/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![feature(arbitrary_self_types)]
+#![feature(let_chains)]
+
+mod sync;
+mod test;
+
+use std::collections::VecDeque;
+use std::fmt::{Debug, Formatter};
+
+use tokio::sync::oneshot::{channel, Receiver, Sender};
+
+use crate::sync::atomic::{AtomicBool, AtomicU64, Ordering as AtomicOrdering};
+use crate::sync::{Arc, Mutex};
+
+type RequestQueue = VecDeque<(Sender<MemoryTracker>, u64)>;
+enum MemoryRequest {
+    Ready(MemoryTracker),
+    Pending(Receiver<MemoryTracker>),
+}
+
+struct MemoryLimiterInner {
+    total_size: AtomicU64,
+    controller: Mutex<RequestQueue>,
+    has_waiter: AtomicBool,
+    quota: u64,
+}
+
+impl MemoryLimiterInner {
+    fn release_quota(&self, quota: u64) {
+        self.total_size.fetch_sub(quota, AtomicOrdering::SeqCst);
+    }
+
+    fn add_memory(&self, quota: u64) {
+        self.total_size.fetch_add(quota, AtomicOrdering::SeqCst);
+    }
+
+    fn may_notify_waiters(self: &Arc<Self>) {
+        // check `has_waiter` to avoid access lock every times drop `MemoryTracker`.
```

Comment:
Only this method is changed compared to the original code in `src/storage/hummock/utils.rs`. 

Changed logic is described as in the PR description.